dockerfile2llb: filter unused paths for named contexts

jsternberg opened the pull request:
This changes the llb generation in `dockerfile2llb` to add the
`llb.FollowPaths` option when named contexts are used in the same way
that happens with the default context.

This means that dockerfiles that look like this:

```
FROM scratch
COPY --from=mynamedctx ./a.txt /a.txt
```

Will only load the file `a.txt`. This behavior is consistent with the
default context.

This also removes the unused []llb.LocalOption from ContextOpt and
replaces it with an async counterpart. This is needed because we
initialize the named context before we know which paths are needed for
the filter.

Fixes #1765.

jsternberg commented:
Want some early feedback on this, but leaving it as a draft as it presently only works when `mynamedctx` is an already defined stage that's being overwritten and doesn't work when it's defined only in the command line. I think I already know where the issue is, but I didn't originally modify that section because I wasn't quite sure what it did.

jedevc commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -756,9 +832,15 @@ func dispatch(d *dispatchState, cmd command, opt dispatchOpt) error {
-		if err == nil && len(cmd.sources) == 0 {
-			for _, src := range c.SourcePaths {
-				d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
+		if err == nil {
+			if len(cmd.sources) == 0 {
+				for _, src := range c.SourcePaths {
+					d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
+				}
+			} else if source := cmd.sources[0]; source.namedCtxState != nil {
+				for _, src := range c.SourcePaths {
+					source.namedCtxState.includePath(src)
+				}
```
I think this logic would need to be replicated to the `instructions.AddCommand` branch as well.

Also, ideally we could somehow just have a single combined struct that wraps both a stage source, and a named context.

jsternberg replied:
I don't think this needs to be replicated to `ADD`. `ADD` doesn't seem to support the `--from` option. The `CopyCommand` has a `From` option while `AddCommand` doesn't seem to have that.

jedevc commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -135,26 +135,86 @@ func ListTargets(ctx context.Context, dt []byte) (*targets.List, error) {
 	return l, nil
 }
 
+type namedContextState struct {
```
Maybe something like `filteredLocalState` or similar?

I'm not sure about including an `Image` field in here, I get the logic, but it's not really connected with what paths get pulled from the client, it shouldn't affect the contents of the image data. If it gets split out, it feels like a separate refactor IMO, so if there's really a good reason to do this, it makes sense as a separate commit.

jedevc commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -135,26 +135,86 @@ func ListTargets(ctx context.Context, dt []byte) (*targets.List, error) {
+	copt := nc.copt
+	if len(nc.includePaths) > 0 {
+		if includePatterns := normalizeContextPaths(nc.includePaths); includePatterns != nil {
+			copt.LocalOpts = append(copt.LocalOpts, llb.FollowPaths(includePatterns))
```
`copt.LocalOpts` will get modified on the original here, which I imagine isn't the intent (given the above comment)?

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -135,26 +135,86 @@ func ListTargets(ctx context.Context, dt []byte) (*targets.List, error) {
+			copt.LocalOpts = append(copt.LocalOpts, llb.FollowPaths(includePatterns))
+		}
+	}
+	st, img, err := client.NamedContext(ctx, nc.name, copt)
```
It would be better to avoid multiple calls to `NamedContext`. Some branches in `NamedContext` do config resolving etc.

What if `dockerui/client.NamedContext()` returns a struct that has `State/Image` properties and `SetLocalOpts([]llb.LocalOption)` method. Current`ContextOpt.LocalOpts` seems to be unused and can be replaced with the new method.

Calling `SetLocalOpts` will change the definition of `State`, either because the internal `llb.Local` is under `mutableOutput` or inside `llb.Async()`. This is only for the `llb.Local` definition though and doesn't rerun any other branches of `NamedContext()`.

jsternberg replied:
I think that could work.

jedevc replied:
If we do that, would it make sense to also return that same struct from the `MainContext` method as well? That way we can use the same way of updating the opts later on.

tonistiigi replied:
One difference is that `MainContext` doesn't ever return an `Image` (at least atm). Otherwise, I think if it is easy and unlikely to break something, we can do it now. Otherwise, we can do the namedcontext for now and look at the case for maincontext in a follow-up.

jsternberg replied:
One thing I was thinking about was what if we just returned it as part of the output similar to `mutableOutput`? Something like:

```
type localOutput struct {
    name string
    opts []llb.LocalOpts
}
```

Then have it create the vertex in an async manner. It should then be possible to inspect the output for the mounted output and update the local options.

I think this could be coded similar to `Async` or possibly coded the way async is coded.

tonistiigi replied:
I'm not sure I understand where/how that `localOutput` would be used.

jsternberg replied:
Borrowing heavily from async:

```
type LocalContext struct {
	name string
	opts []llb.LocalOption
}

func (l *LocalContext) ToInput(ctx context.Context, c *llb.Constraints) (*pb.Input, error) {
	out := l.Do(ctx, c)
	return out.ToInput(ctx, c)
}

func (l *LocalContext) Vertex(ctx context.Context, c *llb.Constraints) llb.Vertex {
	out := l.Do(ctx, c)
	return out.Vertex(ctx, c)
}

func (l *LocalContext) Do(ctx context.Context, c *llb.Constraints) llb.Output {
	st := llb.Local(l.name, l.opts...)
	return st.Output()
}

func main() {
	st1 := llb.NewState(&LocalContext{
		name: "context",
	})

	st2 := st1.Dir("/src")
        // We can modify the local options by inspecting the state output.
	if lc, ok := st2.Output().(*LocalContext); ok {
		lc.opts = append(lc.opts, llb.WithCustomName("internal"))
	}

	def, err := st2.Marshal(context.TODO())
	if err != nil {
		panic(err)
	}
	_ = llb.WriteTo(def, os.Stdout)
}
```

The intention is that the areas where we have to modify the local context, such as this one:

```go
if err == nil && len(cmd.sources) == 0 {
	for _, src := range c.SourcePaths {
		d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
	}
}
```

This could be modified to:

```go
if err == nil {
	if len(cmd.sources) == 0 {
		for _, src := range c.SourcePaths {
			d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
		}
	} else if lc, ok := cmd.sources[0].state.Output().(*LocalContext); ok {
		// see above
	}
}
```

The benefit of this is we could also end up consolidating how the local options are handled so the default context and named contexts aren't different from each other.

I'll play around with the code though. I do have concerns about this being a bit too complicated to follow. Modifying the client to return a struct may end up just being easier.

jsternberg replied:
Decently certain it's possible to do this using `Async` as long as I can modify the values the client returns.

At the moment, the only way to use async for a source is to do `llb.Scratch().Async(...)` and ignore the input. Are we potentially interested in adding an `llb.Async` top-level function? I could use it and can add it as part of this change if that's helpful.

tonistiigi replied:
`llb.Scratch().Async(...)` and `llb.Async` should be equivalent. The previous state (scratch in your case) is just what gets sent to the callback as the previous state value. The callback can just ignore it.

jsternberg replied:
Yea they're equivalent. I was wondering if we wanted to formally add an `llb.Async` that is a shortcut for the above.

jsternberg commented:
I've updated this PR with the feedback. This now uses `Async` to produce the local source. Other paths aren't modified now and the named context isn't created multiple times.

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -778,6 +807,7 @@ type dispatchState struct {
 	buildArgs      []instructions.KeyValuePairOptional
 	commands       []command
 	ctxPaths       map[string]struct{}
+	namedCtxState  *namedContextState
```
I think we can avoid the `namedContextState` type. I would define `includePaths map[string]struct{}` in here and explicitly call `SetLocalOpts` like you had in the previous version to make it clear what is the order of gathering paths and marshaling. If you need a helper function then that is fine but I think `normalizeContextPaths` already does the job.

jsternberg replied:
Yea this makes sense.

tonistiigi commented on frontend/dockerui/namedcontext.go:
```diff
@@ -208,46 +214,64 @@ func (bc *Client) namedContextRecursive(ctx context.Context, name string, nameWi
+		// read from the anonymous function passed to async. In order to ensure we're
+		// referencing the same slice, we allocate a pointer to a slice rather than just
+		// the slice which would be passed by value.
+		customOpts := new([]llb.LocalOption)
```
I don't think this needs to be a pointer. You can just do:
```
var nc NamedContext
st = llb.Async(func() { if nc.localOpts {} })
nc.State = st
return &nc
```

jsternberg replied:
Originally this was going to be a slice that was modified. I think that doesn't really make sense anymore so I'll switch this up. It may be good to make it a bit more explicit though that the pointer is the one being captured. I think go should figure it out automatically, but just in case it would probably be better to do `nc := &NamedContext{}`.

tonistiigi reviewed (commented):
(follow-up) Maybe after https://github.com/tonistiigi/fsutil/pull/167 we can write an integration test for this as well.

jsternberg commented:
Need to do some testing but responded to the existing comments.

jsternberg commented:
Did some testing and I think this is likely good to go. There's on case that I'm concerned about doing incorrectly, which is when the named context is the subject of `FROM` rather than referenced as a mount point, where it shouldn't do any filtering.

Before merging this, it might be beneficial to write an integration test for this that ensures the data transfer is the same between the build context and a named context. I think https://github.com/moby/buildkit/pull/4165 could be beneficial for this since it will be easier to find the number of bytes transferred.

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -382,18 +383,21 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
-						d.state = st.Platform(*platform)
+						d.state = nc.State.Platform(*platform)
 						d.platform = platform
+						// Do not store the named context when it is being used as a base image.
```
But if `d.base` is a `namedContext` already then the filter should also be disabled for it.

One way to do it if you also remove the `namedContexts` map would be to initialize `usedPaths` map on setting `dispatchState.namedContext` and then in here if you detect that it is used as base image then set `usedPaths = nil` what would mean that filtering is disabled.

jsternberg replied:
I'm not completely certain if this comment is even correct or relevant. I need to check on that today. When I was attempting to test it, I ran into a [panic](https://github.com/moby/buildkit/issues/4177) in the scenario that I was concerned about.

Another way of doing this might be to add `"/"` as a used path when I detect it's being used as a base image. Setting it to nil works, but I feel like it's too much state to keep in your head and too many null pointers that could result in a panic. If we add `"/"` as a used path, then `normalizeContextPaths` will return nil and the local option won't be set.

jsternberg replied:
A thing I found to work is disabling the context filtering when the associated dispatch state has commands associated with it. I figure this works well because even if an external stage is using only path `/a.txt`, the stage itself is using the entire base image so I just have commands disabling the filter for named context in general.

When `COPY --from` is used, the implicit stages that are created don't have any commands in them so it's safe to filter them.

tonistiigi reviewed (commented):
With a Dockerfile like
```
from scratch
copy foo .
copy --from=base bar .
```
```
buildctl build --frontend dockerfile.v0 --local dockerfile=. --local context=. --local base=. --opt context:base=local:base
```

this currently does not work.

```
/tmp/a # ls -lh
total 102M
-rw-r--r--    1 root     root          47 Aug 25 05:22 Dockerfile
-rw-r--r--    1 root     root           4 Aug 25 05:21 bar
-rw-r--r--    1 root     root           3 Aug 25 05:21 foo
-rw-r--r--    1 root     root      102.0M Aug 25 05:21 rand

 => [context base] load from client  
 => => transferring base: 106.95MB
``` 

If I add `FROM alpine AS base` as first line that seems to correct it.


Make sure to take this out of draft as well if it is ready.

For the integration test I think https://github.com/tonistiigi/fsutil/pull/167 would be better than trying to read the progress. That would allow defining context as virtual fs where the test can then see what exact files were read from it. Even if we read progress then in the test we can just access the `client` progress types directly and don't need to convert to json and back.

tonistiigi commented:
PTAL CI errors

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -806,12 +833,25 @@ func (dss *dispatchStates) addState(ds *dispatchState) {
+		// Propagate the set of paths from this dispatchState to the base
+		// dispatchState to enable filtering local contexts based on the set
+		// of used paths.
+		ds.paths = mergePaths(ds.paths, d.paths)
```
isn't it a problem that this runs before `dispatchRun`.

I'm not sure how a path added in the dispatchRun of a child stage would affect the dispatchState of the parent that has `namedContext` set.

jsternberg replied:
I think I should probably change the name of this. I didn't think of a better name yesterday.

The intention of this method is to ensure that `ds.paths` and `d.paths` refer to the same location in memory. At this point, `ds.paths` is probably empty but it's just making sure that everything is present in case it wasn't. This happens before `dispatchRun` because it ensures that all of the maps relate to the same map.

I think `unifyPaths` is probably a better name now that I had a night to sleep on it.

jsternberg commented:
Added a test. I think this is good for review now. I would agree that the test for this would be easier if one of two things was true.

1. I could test the underlying LLB output.
2. I could mock the filesystem.

I saw @jedevc had a PR to get that wired up so I can return to this once that's merged. I might also try to experiment with testing underlying LLB structures.

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -6063,6 +6067,116 @@ COPY --from=build /foo /out /
+FROM foo AS image_source
+COPY --from=alpine / /
+RUN cat /foo > /bar
+
```
Could we add a case that also checks the alias case so that the `unifyPaths` gets called.

jsternberg commented:
Updated this PR and it should now pass the tests.

Some changes from the original version:

* `unifyPaths` was removed and the propagation of the paths was also moved. It's now in the same area where the other properties from the base image are propagated in the main loop.
* I was unable to use `llb.Async` because `llb.Async` was triggered when a later section used `AddEnv`. Internally, `AddEnv` calls `getEnv` which forces the state to be initialized. When this happened, it was before the necessary section of code was run to allow `Async` to work properly in this context.
* `NamedContext` was reverted back to returning `(*llb.State, *image.Image, error)`. I found that returning `*NamedContext` created too tight of a binding between the `dockerfile2llb` and `dockerui` packages in a way that was really only relevant to one of the two packages. It ended up being easier to change `LocalOpts` in `ContextOpt` to `AsyncLocalOpts` and build the async logic into `dockerui`.
* The target now always has `/` included as a necessary path.
